Node power cap support for Intel Broadwell

tpatki opened the pull request:
Add support for power capping a node on Intel Broadwell. As discussed, we take the input value and divide it by number of sockets, and set this as the uniform power cap on each socket. We will develop a model later that accomodates for memory, uncore and other components. This is a first cut to add support for Flux on Quartz.

tpatki committed:
Add support for power capping a node on Intel Broadwell. We take the input value and divide it by number of sockets, and set uniform power caps on each socket. We will develop a model later that accomodates for memory, uncore and other components. This is a first cut to add support for Flux on Quartz.

rountree commented:
Calling `fm_06_4f_set_node_power_limit()` doesn't set a node power limit on Broadwell, and I don't want to be explaining to users for the next few years why that's the case.

I would prefer the cross-architectural function be `set_cappable_power_limit()` and leave the idea of "node" out of the equation entirely.

tpatki committed:
Minor text edit

tpatki commented:
Fair enough. If I understand this correctly, you are expecting the lower-level function name to reflect this, correct? And not the top-level API (that is, variorum_set_node_power_limit)? The top-level needs to be vendor neutral, and we can rename that as well (on both IBM and Intel).

tpatki commented:
We decided to go with `variorum_set_best_effort_node_power_limit` as a vendor neutral option that is exposed to external tools. I'll update this PR shortly for both IBM and Intel implementations.

tpatki commented:
Astyle checks keep failing to format if there are more than 80 chars, which was the case in several places for variorum error handlers. I'm hoping the current commit fixes that as well...

tpatki committed:
Changes API from variorum_set_node_power_limit to variorum_set_best_effort_node_power_limit.

Modifies IBM and Intel implementations to reflect this.
For IBM systems, a node power cap is available in hardware and is used directly.
On Intel Broadwell, we have best effort capping.
We take the input value and divide it by number of sockets,
and set this as the uniform power cap on each socket.
We will develop a model later that accomodates for memory, uncore and other components in the future.
This is a first cut to add support for Flux on Quartz and Lassen.
Has been tested on both systems.

tpatki commented:
@slabasan @rountree 
This is ready for review. Tested on Quartz and Lassen. 

I also made some minor formatting changes (limiting to 80 chars per row) that seem to always cause issues with astyle when when those segments of code are not edited.

slabasan commented on src/variorum/Intel/Broadwell_4F.c:
```diff
@@ -411,3 +426,35 @@ int fm_06_4f_get_node_power_json(json_t *get_power_obj)
+    variorum_get_topology(&nsockets, &ncores, &nthreads);
+
+    // Adding this for portability and rounding down.
+    // Ideally line 437 should be okay as it is integer division and we have
```
Line 437 here is referencing a comment. Did you mean line 451?

slabasan commented on src/examples/variorum-set-best-effort-node-power-limit-example.c:
```diff
@@ -15,17 +15,17 @@ int main(int argc, char **argv)
-        printf("No power limit specified...using default limit of 500W.\n");
-        // 500W is based on minimum power on IBM Witherspoon
-        node_pow_lim_watts = 500;
+        printf("Please specify an input value in W for correctness.\n");
```
Spell out Watts here instead of just W.

tpatki committed:
Address comments from Stephanie's review

tpatki commented:
@slabasan Just a reminder on this.

slabasan reviewed (commented):
+1

Also just noting here that this PR includes manual code formatting. Automatic char width in astyle counts the number of chars with white space at the beginning of the line (i.e., tabs) removed. Because of this, lines can contain more than 80 chars (including the white space). Unfortunately, there is no configuration in astyle to include whitespace in the width calculation.
